File: python3/day_11_Sort Colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        left = -1
        right = len(nums)
        i = 0
        while i < right:
            if left + 1 < right and nums[left+1] == 0:
                left += 1
                i += 1
                continue
            
            if right - 1 > left and nums[right-1] == 2:
                right -= 1
                continue

            
            if nums[i] == 0:
                left += 1
                nums[left], nums[i] = nums[i], nums[left]
            elif nums[i] == 2:
                right -= 1
                nums[right], nums[i] = nums[i], nums[right]
            else:
                i += 1

File: cpp/june/day_11_Sort Colors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int left, right, i;
        left = -1;
        right = nums.size();
        i = 0;
        while(i < right){
            if(left+1 < right && nums[left+1] == 0){
                left++;
                i++;
                continue;
            }
            if(right > 1 && nums[right-1] == 2){
                right--;
                continue;
            }
            if(nums[i] == 0){
                left++;
                swap(nums[i], nums[left]);
            }else if(nums[i] == 2){
                right--;
                swap(nums[i], nums[right]);
            }else{
                i++;
            }
        }
    }
};
